### PR TITLE
feat(typescript): add ts_project rule

### PR DESCRIPTION
**Pull request by alexeagle**
This is a very thin layer on top of vanilla tsc

### CODE REVIEW

File: examples/react_webpack/BUILD.bazel [hunk 1]
```diff
@@ -13,22 +13,14 @@ sass(
...
-        "react",
-    ],
-    data = [
+ts_project(
+    name = "tsconfig",
+    srcs = [
         "index.tsx",
         "types.d.ts",
+    ],
+    declaration = False,
```
Comment (mrmeku) [inline examples/react_webpack/BUILD.bazel]: Can you a comment describing why declaration is set to false?

File: packages/typescript/src/internal/ts_project.bzl [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    # ts, tsx, js, jsx, json
+    "srcs": attr.label_list(allow_files = True),
+    "extends": attr.label_list(allow_files = [".json"]),
+    "tsc": attr.label(default = Label("@npm//typescript/bin:tsc"), executable = True, cfg = "host"),
+    "tsconfig": attr.label(mandatory = True, allow_single_file = [".json"]),
+    "deps": attr.label_list(providers = [DeclarationInfo]),
+}
+
+_OUTPUTS = {
+    "js_outs": attr.output_list(),
```
Comment (mrmeku) [inline packages/typescript/src/internal/ts_project.bzl]: Could you maybe address the difference in outputs of ts_projecrt and ts_library in a comment or readme file somewhere in the PR?

File: packages/typescript/src/internal/ts_project.bzl [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    run_node(
+        ctx,
+        inputs = inputs,
+        arguments = [arguments],
+        outputs = outputs,
+        executable = "tsc",
+        progress_message = "Compiling TypeScript project %s" % ctx.file.tsconfig.short_path,
+    )
+
+    runtime_files = depset(ctx.outputs.js_outs + ctx.outputs.map_outs)
```
Comment (mrmeku) [inline packages/typescript/src/internal/ts_project.bzl]: I'm not sure if this is important by if a src file is a `.js` dependency might it also need to be included as a runtime file?

Comment (Toxicable) [reply packages/typescript/src/internal/ts_project.bzl]: If you have .js inputs, does tsc write them to the our dir? or does it only emit diagnostics?

File: packages/typescript/src/internal/ts_project.bzl [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            Instructs Bazel to expect a `.js.map` output for each `.ts` source.
+        declaration_map: if the `declarationMap` bit is set in the tsconfig.
+            Instructs Bazel to expect a `.d.ts.map` output for each `.ts` source.
+        emit_declaration_only: if the `emitDeclarationOnly` bit is set in the tsconfig.
+            Instructs Bazel *not* to expect `.js` outputs for `.ts` sources.
+    """
+
+    if srcs == None:
+        srcs = native.glob(["**/*.ts"])
+    tsconfig = name + ".json"
```
Comment (Toxicable) [inline packages/typescript/src/internal/ts_project.bzl]: Not sure if im a fan of this.
Could we possibly change the logic to default to `tsconfig = ctx.testonly ? 'tsconfig-test.json' : 'tsconfig.json'`.
But override with the tsconfig speciffied by the user if exists.
Usually we'd have libs where their "primary target" is a `ts_library` and this will use the same name as the dir, making it easier to reference.
This here prevents using a name that's different from your tsconfig, right?
Is there another reason why this might be a good idea?

File: packages/typescript/src/internal/ts_project.bzl [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+def ts_project_macro(
+        name = "tsconfig",
+        srcs = None,
+        deps = [],
+        extends = None,
+        declaration = True,
+        source_map = False,
+        declaration_map = False,
+        emit_declaration_only = False,
```
Comment (Toxicable) [inline packages/typescript/src/internal/ts_project.bzl]: maybe instead have a flag that says `emit_js = True`? then this can be derived from the inverse of that flag?

File: packages/typescript/src/internal/ts_project.bzl [hunk 6]
```diff
@@ -0,0 +1,192 @@
+"ts_project rule"
+
+load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo", "NpmPackageInfo", "run_node")
+
+_ATTRS = {
+    # TODO: limit to the extensions we expect?
+    # ts, tsx, js, jsx, json
+    "srcs": attr.label_list(allow_files = True),
+    "extends": attr.label_list(allow_files = [".json"]),
+    "tsc": attr.label(default = Label("@npm//typescript/bin:tsc"), executable = True, cfg = "host"),
```
Comment (Toxicable) [inline packages/typescript/src/internal/ts_project.bzl]: i'd love to see if this works by replacing this with `ngtsc`

File: examples/react_webpack/BUILD.bazel (hunk 1 above)
Comment (Toxicable) [reply examples/react_webpack/BUILD.bazel]: Looks like it's a demo showing a 1:1 with the old `tsc`  which didn't do declarations

Comment (alexeagle) [reply examples/react_webpack/BUILD.bazel]: added a comment, yes this is exercising the codepath where tsc wasn't configured with `declaration` or `composite`

File: packages/typescript/src/internal/ts_project.bzl (hunk 2 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: Ultimately I think that needs to go in packages/typescript/docs/install.md where there's currently a discussion of `ts_library` vs `@npm//typescript:index.bzl#tsc`

I'll add a comment here about just mirroring what tsc can produce

File: packages/typescript/src/internal/ts_project.bzl (hunk 3 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: yeah this needs a test case for .js inputs...

File: packages/typescript/src/internal/ts_project.bzl (hunk 2 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: and a TODO about updating the markdown docs after this lands

File: packages/typescript/src/internal/ts_project.bzl (hunk 4 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: I'll add a comment in the code explaining this.

```
Bazel's `name` property is arbitrarily controlled by the user. But here we constrain it to match the tsconfig file.
The reason for doing this is so that BUILD file generation can be trivial. If you know the path of a referenced tsconfig file, you know what label to use to include it in the deps of another rule.
If we allowed users to control it, then BUILD file generation would need heuristic semantics to do bazel query and try to figure out which label gives a target that is usable as a dep.
```

And on the rule docs
```
You can use a Bazel alias() to give an additional label for this typescript project compilation
```

Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: Also note you could import the raw `ts_project` rule from the `internal` path, which I think we should leave undocumented, but it gives you the escape valve to avoid this macro logic.

File: packages/typescript/src/internal/ts_project.bzl (hunk 5 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: I'm trying to keep these flags named identically to the tsconfig file so the mapping between the two is totally obvious.

File: packages/typescript/src/internal/ts_project.bzl (hunk 6 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: yeah I played with it briefly and something didn't work, please give that a try if you have time :)

File: packages/typescript/src/internal/ts_project.bzl (hunk 5 above)
Comment (Toxicable) [reply packages/typescript/src/internal/ts_project.bzl]: Ah right, I wasn't aware of a flag in the tsconfig of the same name, sounds good!

File: packages/typescript/src/internal/ts_project.bzl (hunk 4 above)
Comment (Toxicable) [reply packages/typescript/src/internal/ts_project.bzl]: Ah yeah I see the constraint here, sounds good to me with the extra docs and code gen considered

File: examples/react_webpack/tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{
+    "compilerOptions": {
+        "jsx": "react",
+        "lib": ["ES2015", "DOM"],
+        "listEmittedFiles": true,
+        "traceResolution": true
```
Comment (Toxicable) [inline examples/react_webpack/tsconfig.json]: Do these two options need to be enabled?

File: packages/typescript/src/internal/ts_project.bzl [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+#    cd C:/b/uuxnwop3/execroot/build_bazel_rules_nodejs
+#    SET COMPILATION_MODE=fastbuild
+#    bazel-out/host/bin/external/npm/typescript/bin/tsc.bat -p packages/typescript/test/ts_project/b/tsconfig.json --outDir bazel-out/x64_windows-fastbuild/bin/packages/typescript/test/ts_project/b --declarationDir bazel-out/x64_windows-fastbuild/bin/packages/typescript/test/ts_project/b --bazel_node_modules_manifest=bazel-out/x64_windows-fastbuild/bin/packages/typescript/test/ts_project/b/_tsconfig.module_mappings.json --nobazel_patch_module_resolver
+#  Execution platform: @local_config_platform//:host
+#  packages/typescript/test/ts_project/b/b.ts(1,17): error TS6305: Output file 'C:/b/uuxnwop3/execroot/build_bazel_rules_nodejs/packages/typescript/test/ts_project/a/a.d.ts' has not been built from source file 'C:/b/uuxnwop3/execroot/build_bazel_rules_nodejs/packages/typescript/test/ts_project/a/a.ts'.
+TS_37378_DISABLED = ["fix-windows"]
+
+_ATTRS = {
+    # TODO: limit to the extensions we expect? ts, tsx, js, jsx, json
+    "srcs": attr.label_list(allow_files = True, mandatory = True),
```
Comment (Toxicable) [inline packages/typescript/src/internal/ts_project.bzl]: I don't think we should limit the extensions here, especically if you can swap out the compiler for `ngtsc` for example, since the inputs would then include .html and .css

File: examples/react_webpack/tsconfig.json (hunk 7 above)
Comment (alexeagle) [reply examples/react_webpack/tsconfig.json]: no, these are for debugging.
I think maybe the rule implementation could enable these when running with VERBOSE_LOGS in the environment...

File: packages/typescript/src/internal/ts_project.bzl (hunk 8 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: yeah I think even tsc keeps adding ability to natively type-check against require('./funny.thing')

File: packages/typescript/src/internal/ts_project.bzl [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            "--listFiles",
+            # Did tsc write all outputs to the place we expect to find them?
+            "--listEmittedFiles",
+            # Why did module resolution fail?
+            "--traceResolution",
+            # Why was the build slow?
+            "--diagnostics",
+            "--extendedDiagnostics",
+            # Prefer to show gory details rather than save space in console
+            "--noErrorTruncation",
```
Comment (evmar) [inline packages/typescript/src/internal/ts_project.bzl]: We recently discovered that tsc can OOM when you set this flag, because without error truncation error generation can infinite loop.

Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: good tip, thanks. I only set this here from reading through the tsc docs and it sounded helpful, will remove

Comment (evmar) [reply packages/typescript/src/internal/ts_project.bzl]: Oh wait they fixed it by actually limiting it with the flag, heh https://github.com/microsoft/TypeScript/pull/37461

File: packages/typescript/src/internal/ts_project.bzl (hunk 4 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: Discussed with @gregmagolan - even if users can control `tsconfig` and `name` attributes independently, it's still pretty easy for the generator to print all `ts_project` rules in the referenced package to find the right one. So `tsconfig` is now independent (though we recommend naming it the basename of the tsconfig.json and that default is convenient)

File: examples/react_webpack/BUILD.bazel [hunk 10]
```diff
@@ -13,27 +13,29 @@ sass(
...
+    deps = [
         "@npm//@types",
         "@npm//csstype",
     ],
 )
 
+# If you don't want to write //:tsconfig as the label for the TypeScript compilation,
+# use an alias like this, so you can write //:compile_ts instead.
+alias(
+    name = "compile_ts",
```
Comment (gregmagolan) [inline examples/react_webpack/BUILD.bazel]: nit: use tsconfig attribute and use "compile" as ts_project name to avoid alias?

File: packages/typescript/src/internal/ts_project.bzl [hunk 11]
```diff
@@ -0,0 +1,240 @@
+"ts_project rule"
+
+load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo", "NpmPackageInfo", "run_node")
+
+_ATTRS = {
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: I think we should add "args" in general to rules to allow users to pass their own arguments as
1) config files do not always support all arguments. I recently added "args" to rollup for this reason so 
Angular could pass `--silent` which is command line one.
2) users may want to override config file settings with command line arguments in some cases
General pattern is
```
    "args": attr.string_list(
        doc = """Command line arguments to pass to <tool name>. Can be used to override config file settings.
```
and 
```
```
    # See CLI documentation at https://rollupjs.org/guide/en/#command-line-reference
    args = ctx.actions.args()

    # Add user specified arguments *before* rule supplied arguments
    args.add_all(ctx.attr.args)
```
where `args` is passed to `run_node`

File: packages/typescript/src/internal/ts_project.bzl [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    # NB: no restriction on extensions here, because tsc sometimes adds type-check support
+    # for more file kinds (like require('some.json')) and also
+    # if you swap out the `compiler` attribute (like with ngtsc)
+    # that compiler might allow more sources than tsc does.
+    "srcs": attr.label_list(allow_files = True, mandatory = True),
+    "extends": attr.label_list(allow_files = [".json"]),
+    "tsc": attr.label(default = Label("@npm//typescript/bin:tsc"), executable = True, cfg = "host"),
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: nit: set DEFAULT_COMPILER = "@npm//typescript/bin:tsc" as the string is in two places

File: examples/react_webpack/BUILD.bazel (hunk 10 above)
Comment (alexeagle) [reply examples/react_webpack/BUILD.bazel]: oops, this BUILD file should have been updated for the change of `declaration` default to False too. will boil it down to the minimum

File: packages/typescript/src/internal/ts_project.bzl [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+
+            Must include any tsconfig files "chained" by extends clauses.
+
+        tsc: Label of the TypeScript compiler binary to run.
+
+            Override this if your npm_install or yarn_install isn't named "npm"
+            For example, `tsc = "@my_deps//typescript/bin:tsc"`
+            Or you can pass a custom compiler binary instead.
+
+        declaration: if the `declaration` bit is set in the tsconfig.
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: nit: links to typescript docs for these 1:1 mappings for more info?

File: packages/typescript/src/internal/ts_project.bzl [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+
+    return [
+        DeclarationInfo(
+            declarations = depset(typings_files),
+            transitive_declarations = depset(typings_files, transitive = [
+                dep[DeclarationInfo].transitive_declarations
+                for dep in ctx.attr.deps
+            ]),
+        ),
+        DefaultInfo(
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: nit: short comment on rationale for only including runtime files in DefaultInfo?

File: packages/typescript/src/internal/ts_project.bzl [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+
+    deps_depsets = []
+    for dep in ctx.attr.deps:
+        if _TsConfigInfo in dep:
+            deps_depsets.append(dep[_TsConfigInfo].tsconfigs)
+        if NpmPackageInfo in dep:
+            # TODO: we could maybe filter these to be tsconfig.json or *.d.ts only
+            # we don't expect tsc wants to read any other files from npm packages.
+            deps_depsets.append(dep[NpmPackageInfo].sources)
+        elif DeclarationInfo in dep:
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: should this just be an `if`? They seem orthogonal and should overlap anyway.

File: packages/typescript/src/internal/ts_project.bzl [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+
+    inputs = ctx.files.srcs + depset(transitive = deps_depsets).to_list() + [ctx.file.tsconfig]
+    if ctx.attr.extends:
+        inputs.extend(ctx.files.extends)
+    outputs = ctx.outputs.js_outs + ctx.outputs.map_outs + ctx.outputs.typings_outs + ctx.outputs.typing_maps_outs
+    if ctx.outputs.buildinfo_out:
+        outputs.append(ctx.outputs.buildinfo_out)
+
+    if len(outputs) == 0:
+        return []
```
Comment (gregmagolan) [inline packages/typescript/src/internal/ts_project.bzl]: should still return transitive declarations & tsconfigs?

File: packages/typescript/src/internal/ts_project.bzl (hunk 11 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: added `args` pass-through with a delightful test coverage

File: packages/typescript/src/internal/ts_project.bzl (hunk 13 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: added above

File: packages/typescript/src/internal/ts_project.bzl (hunk 14 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: # DefaultInfo is what you see on the command-line for a built library,
        # and determines what files are used by a simple non-provider-aware
        # downstream library.
        # Only the JavaScript outputs are intended for use in non-TS-aware
        # dependents.

File: packages/typescript/src/internal/ts_project.bzl (hunk 15 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: hmm, I think this is true. Need to verify if we have coverage for it

File: packages/typescript/src/internal/ts_project.bzl (hunk 16 above)
Comment (alexeagle) [reply packages/typescript/src/internal/ts_project.bzl]: yes, took a while to add a test case to expose this...

File: packages/typescript/src/internal/ts_project.bzl (hunk 14 above)
Comment (gregmagolan) [reply packages/typescript/src/internal/ts_project.bzl]: 👍
